### PR TITLE
♻️ Replace aiopg in catalog service

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?

This PR is following #2864 that updated SQLalchemy to the latest version and left the catalog service with some functional but suboptimal code such that aiopg could work with the newer SQLalchemy syntax.

This PR replaces aiopg/psycopg2 with asyncpg and allows a full integration with SQLalchemy **ONLY** in the catalog for now.
This PR also introduces the use of the library aiocache and is currently used to cache the list of services returned by the director-v0 for a small performance boost in listing the services according to service access rights. This library can also be used with redis to have a distributed cache.
Currently we list all the services with all the details, etc. The most time consuming part is creating the ```ServiceOut``` objects at the end. There I don't see how to fix that for now, unless we reduce the number of objects?

Using the new async engine in SQLalchemy implies:
- replace any occurence of ```acquire()``` with ```connect()```
- if using a construction with ```async for row...``` then one needs to call ```conn.stream(``` instead of ```conn.execute(```
- SQL calls with a string is not allowed, one needs to pass a ```sa.DDL``` type with the string inside

Additional:
- added a fixture to create a SQLalchemy async engine in ```postgres_service.py```
- catalog now caches the list of services in the director-v0 for 5 minutes using a TTL cache

Bonus:
- added pytest-benchmark for benchmarking tests

reference for SQLAlchemy in async mode: [https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#synopsis-core](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#synopsis-core)

### CODE REVIEW

**Review by GitHK — APPROVED** [review]
👍 very nice

File: packages/pytest-simcore/src/pytest_simcore/postgres_service.py [hunk 1]
```diff
@@ -181,6 +180,22 @@ async def aiopg_engine(
...
 
+@pytest.fixture(scope="function")
+async def sqlalchemy_async_engine(
+    postgres_db: sa.engine.Engine,
+) -> AsyncIterator:
+    # NOTE: prevent having to import this if latest sqlalchemy not installed (temp)
+    from sqlalchemy.ext.asyncio import create_async_engine
+
+    engine = create_async_engine(
+        f"{postgres_db.url}".replace("postgresql", "postgresql+asyncpg")
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/pytest-simcore/src/pytest_simcore/postgres_service.py:191]: this is the only location where the async engine is created (pytest-simcore package). but no worries, the PostgresSettings also have a special attribute that does it for you.

File: packages/postgres-database/src/simcore_postgres_database/utils_aiosqlalchemy.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+async def get_pg_engine_stateinfo(engine: AsyncEngine, db_name: str) -> Dict[str, Any]:
+    return {
+        "pgserver stats": f"{await _get_connections(engine, db_name)}",
+        "current pool connections": f"{engine.pool.checkedin()=},{engine.pool.checkedout()=}",
+    }
+
+
+async def close_engine(engine: AsyncEngine) -> None:
+    await engine.dispose()
```
Comment (pcrespov) [inline packages/postgres-database/src/simcore_postgres_database/utils_aiosqlalchemy.py:37]: Does it really worth having a helper for something so simple? 
Remember that increasing the level of indirections and "new interfaces" has a cost !

File: packages/postgres-database/src/simcore_postgres_database/utils_aiosqlalchemy.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+
+class DBMigrationError(RuntimeError):
```
Comment (pcrespov) [inline packages/postgres-database/src/simcore_postgres_database/utils_aiosqlalchemy.py:40]: All exceptions that can be handled by external plugins/libraries ... shall be placed in the corresponding exceptions model. In this case we have ``simcore_postgres_database.errors``

File: packages/postgres-database/src/simcore_postgres_database/utils_aiosqlalchemy.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    pass
+
+
+async def raise_if_migration_not_ready(engine: AsyncEngine):
+    """Ensures db migration is complete
+
+    :raises DBMigrationError
+    :raises
```
Comment (pcrespov) [inline packages/postgres-database/src/simcore_postgres_database/utils_aiosqlalchemy.py:48]: MINOR: rm comment

File: packages/pytest-simcore/src/pytest_simcore/postgres_service.py [hunk 5]
```diff
@@ -181,6 +180,22 @@ async def aiopg_engine(
         await engine.wait_closed()
 
 
+@pytest.fixture(scope="function")
+async def sqlalchemy_async_engine(
+    postgres_db: sa.engine.Engine,
+) -> AsyncIterator:
+    # NOTE: prevent having to import this if latest sqlalchemy not installed (temp)
```
Comment (pcrespov) [inline packages/pytest-simcore/src/pytest_simcore/postgres_service.py:187]: if "temp" please add a TODO:

File: services/catalog/requirements/_base.in [hunk 6]
```diff
@@ -21,12 +21,14 @@ fastapi[all]
 pydantic[dotenv]
 
 # database
-aiopg[sa]
+asyncpg
+sqlalchemy[asyncio]
```
Comment (pcrespov) [inline services/catalog/requirements/_base.in:25]: two things, we already introduce this "extra"
```
sqlalchemy[postgresql_psycopg2binary]
```
should we always extra with both?
```
sqlalchemy[asyncio,postgresql_psycopg2binary]
```

File: services/catalog/tests/unit/with_dbs/test_entrypoint_services.py [hunk 7]
```diff
@@ -32,146 +25,214 @@
...
+                product=target_product,
+            )
+            for s in range(NUM_SERVICES)
+        ]
+    )
 
-    async def return_gid_from_email(*args, **kwargs) -> Optional[PositiveInt]:
-        return user_groups[0].gid
+    url = URL("/v0/services").with_query({"user_id": user_id, "details": "false"})
+    response = benchmark(
```
Comment (pcrespov) [inline services/catalog/tests/unit/with_dbs/test_entrypoint_services.py:123]: VERY NICE the benchmark!

File: services/catalog/tests/unit/with_dbs/test_entrypoint_services.py [hunk 8]
```diff
@@ -32,146 +25,214 @@
...
+        client.get, f"{url}", headers={"x-simcore-products-name": target_product}
+    )
+    assert response.status_code == 200
+    data = response.json()
+    assert len(data) == NUM_SERVICES
+    for service in data:
+        assert service["key"] == SERVICE_KEY
+        assert re.match("1.0.[0-9]+", service["version"]) is not None
```
Comment (pcrespov) [inline services/catalog/tests/unit/with_dbs/test_entrypoint_services.py:131]: FYI: there is a regex for version in ``models_library.basic_regex``

File: tests/performance/locust_files/catalog_servicesd.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
+#
+# SEE https://docs.locust.io/en/stable/quickstart.html
+#
+
```
Comment (pcrespov) [inline tests/performance/locust_files/catalog_servicesd.py:4]: MINOR: why ending in "*d.py" the filename?

**Review by pcrespov — APPROVED** [review]
Very nice work. Just a few comments!

File: packages/postgres-database/src/simcore_postgres_database/utils_aiosqlalchemy.py (hunk 2 above)
Comment (sanderegg) [reply packages/postgres-database/src/simcore_postgres_database/utils_aiosqlalchemy.py:37]: true that. I actually did a copy of the utils_aiopg.py file... but I agree this is not necessary

File: services/catalog/requirements/_base.in (hunk 6 above)
Comment (sanderegg) [reply services/catalog/requirements/_base.in:25]: no I think once we can fully migrate, the psycopg2binary becomes obsolete. maybe just for testing where we do have a sync engine as well

File: services/catalog/tests/unit/with_dbs/test_entrypoint_services.py (hunk 7 above)
Comment (sanderegg) [reply services/catalog/tests/unit/with_dbs/test_entrypoint_services.py:123]: Thanks! I hoped someone would see it!

File: tests/performance/locust_files/catalog_servicesd.py (hunk 9 above)
Comment (sanderegg) [reply tests/performance/locust_files/catalog_servicesd.py:4]: oups... d for direct isn't that obvious? :rofl: will change it

File: packages/postgres-database/src/simcore_postgres_database/utils_aiosqlalchemy.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+import sqlalchemy as sa
+from sqlalchemy.ext.asyncio import AsyncEngine
+
+from .utils_migration import get_current_head
+
+
+async def _get_connections(engine, db_name: str) -> List[Dict]:
+    """Return information about connections"""
+    sql = sa.DDL(
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/postgres-database/src/simcore_postgres_database/utils_aiosqlalchemy.py:11]: no idea. I remove it

**Pull request merged**